Add tutorial for blob storage (#18)
MS-SJF-Projects/web-game-complete: modules/05_add_blob_storage/README.md

# Add persistent storage for images

## Goal

Modify the app to add persistent storage for images.

## Tutorial

> In this tutorial we will assume that you've completed the previous tutorial
and already have a game where you get an image and you have to guess a word that describes it.

> Now, we'll need to change the image storing functionality and upload the images to the cloud.

Reviewer: ```suggestion

Now, we'll need to change the image storing functionality and upload the images to the cloud.
```

> Then images could be served from the cloud.

Reviewer: ```suggestion
Then images will be served from the cloud and won't take precious memory space on our server.
```

### Part 1. Create the storage account

#### Intro

[Azure Blob Storage](https://docs.microsoft.com/azure/storage/blobs/) is Microsoft's object storage solution

>  for the cloud. We will use this to store our image and also to get them from there using an URL. It's advisable that

Reviewer: ```suggestion
 for the cloud. We will use it to store our image and also to get them from there using a URL. It's advisable that
```

 you read about this service and familiarise yourself with the concepts of a storage account, container and blob.

#### 1. Create a storage account in your own subscription where you will save the images

In order to store the images, we recommend that you create your own storage account using the portal.
The required steps are provided [here](https://docs.microsoft.com/azure/storage/common/storage-account-create?tabs=azure-portal)

#### 2. Save and expose the storage account information

In order to connect to the storage account from our code, we will need the name of the storage account and a
 storage account key that can be found in the Access Keys of the storage account like shown
 [here](https://docs.microsoft.com/azure/storage/blobs/storage-quickstart-blobs-python#copy-your-credentials-from-the-azure-portal).
Don't copy the connection string, copy the key instead, as we will use this to build the connection string
 and also to get the URL for each uploaded file.
These should be added to two environment variables: AZURE_STORAGE_ACCOUNT_NAME and AZURE_STORAGE_ACCOUNT_KEY.

> An example on how to add an environment variable is [here](https://docs.microsoft.com/azure/storage/blobs/storage-quickstart-blobs-python#configure-your-storage-connection-string).

Reviewer: I think this particular link isn't helpful. They'll rather need an instruction how to set env variables in vscode .

> #### Modify the application code to store data into the storage account

Reviewer: ```suggestion
#### Modify the application code to store data in the storage account
```

Information on how to manage blobs with a storage account using Python is found [here](https://docs.microsoft.com/azure/storage/blobs/storage-quickstart-blobs-python)
We first need to add azure-storage-blob to the requirements.txt file and install the new dependency by using pip:
`pip install -r requirements.txt`.

In the `in_memory_storage.py`, modify the StorageItem.
We can replace the image_bytes and the image_content_type with an image_url that will link to the image saved in the storage account.
Now the object will look like this:

```python
class StorageItem:
    image_url: str
    secret_word: str
```

Don't forget to modify the unit tests to use this new structure.

When we upload the image, in `def upload_image()` we will add the code that uploads the image in the storage account.

> You can create a new class blob_storage.py that will define how we upload the image.

Reviewer: ```suggestion
Create a new `BlobStorage` class in a new file blob_storage.py. This class we'll use to upload images to blob storage.
```
